ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandRunner.java

```diff
@@ -127,38 +127,55 @@ public void processPriorCommands() {
       terminateCluster(terminateCmd.get().getCommand());
       return;
     }
+
     restoreCommands.forEach(
-        command -> RetryUtil.retryWithBackoff(
-            maxRetries,
-            STATEMENT_RETRY_MS,
-            MAX_STATEMENT_RETRY_MS,
-            () -> statementExecutor.handleRestore(command),
-            WakeupException.class
-        )
+        queuedCommand -> {
+          final int topicOffset = queuedCommand.getCommand().getCommandTopicOffset();
+          if (topicOffset == -1
+              || topicOffset == commandStore.getSnapshotWithOffset().getSnapshotOffset()) {
+            commandStore.setOffsetValue(queuedCommand.getOffset() + 1);
+            RetryUtil.retryWithBackoff(
+                maxRetries,
+                STATEMENT_RETRY_MS,
+                MAX_STATEMENT_RETRY_MS,
+                () -> statementExecutor.handleRestore(queuedCommand),
+                WakeupException.class
+            );
+            commandStore.setSnapshot();
+          }
+        }
     );
     final KsqlEngine ksqlEngine = statementExecutor.getKsqlEngine();
     ksqlEngine.getPersistentQueries().forEach(PersistentQueryMetadata::start);
   }
 
   void fetchAndRunCommands() {
-    final List<QueuedCommand> commands = commandStore.getNewCommands(NEW_CMDS_TIMEOUT);
-    if (commands.isEmpty()) {
+    final List<QueuedCommand> queuedCommands = commandStore.getNewCommands(NEW_CMDS_TIMEOUT);
+
+    if (queuedCommands.isEmpty()) {
       return;
     }
 
-    final Optional<QueuedCommand> terminateCmd = findTerminateCommand(commands);
+    final Optional<QueuedCommand> terminateCmd = findTerminateCommand(queuedCommands);
     if (terminateCmd.isPresent()) {
       terminateCluster(terminateCmd.get().getCommand());
       return;
     }
 
-    log.trace("Found {} new writes to command topic", commands.size());
-    for (final QueuedCommand command : commands) {
+    log.trace("Found {} new writes to command topic", queuedCommands.size());
+
+    for (final QueuedCommand queuedCommand : queuedCommands) {
       if (closed) {
         return;
       }
 
-      executeStatement(command);
+      final int commandTopicOffset = queuedCommand.getCommand().getCommandTopicOffset();
+      if (commandTopicOffset == -1
+          || commandTopicOffset == commandStore.getSnapshotWithOffset().getSnapshotOffset()) {
+        commandStore.setOffsetValue(queuedCommand.getOffset() + 1);
```

Comment:
I ended up splitting updating the offset and the snapshot into two separate calls because there appears to be a race condition that can occur when a KsqlRequest contains multiple statements.

The integration test KsqlResourceFunctionalTest.shouldHandleInterDependantCsasTerminateAndDrop was failing

```
 final List<KsqlEntity> results = makeKsqlRequest(
        "CREATE STREAM SS AS SELECT * FROM " + PAGE_VIEW_STREAM + ";"
            + "TERMINATE CSAS_SS_" + NEXT_QUERY_ID.get() + ";"
            + "DROP STREAM SS;"
```

These commands are being pushed to the command topic in this order
```
Command{statement='CREATE STREAM SS WITH (KAFKA_TOPIC='SS', PARTITIONS=1, REPLICAS=1) AS SELECT *
FROM PAGEVIEWS_ORIGINAL PAGEVIEWS_ORIGINAL
EMIT CHANGES;',commandTopicOffset=1, overwriteProperties={}}
Command{statement='TERMINATE CSAS_SS_0;',commandTopicOffset=1, overwriteProperties={}}
Command{statement='DROP STREAM SS;',commandTopicOffset=2, overwriteProperties={}}
```
The terminate command isn't being executed since the offset signature doesn't match the current one, which then causes the drop statement to fail also.

Reply:
Another option, which I think I like better would be to add the notion of a "batch" to `CommandStore`. So you could have something like:

```
class CommandStore {
    BatchContext newBatch(final long offset) {
        return new BatchContext(offset);
    }

    // then add BatchContext as an argument
    enqueueCommand(final BatchContext ctx, final String statement, ...) {
        final Command command = new Command(ctx.offset(),...);
        ctx.incrementOffset();
    }
}
```